lightning/src/ln/channel.rs

```diff
@@ -1160,6 +1160,45 @@ impl<'a, SP: Deref> ChannelPhase<SP> where
 	}
 }
 
+/// A top-level channel struct, containing a channel phase
+pub(super) struct ChannelWrapper<SP: Deref> where SP::Target: SignerProvider {
+	phase: ChannelPhase<SP>,
+}
+
+impl<'a, SP: Deref> ChannelWrapper<SP> where SP::Target: SignerProvider {
+	pub fn new(phase: ChannelPhase<SP>) -> Self {
+		Self { phase }
+	}
+
+	pub fn phase(&self) -> &ChannelPhase<SP> {
+		&self.phase
+	}
+
+	pub fn phase_mut(&mut self) -> &mut ChannelPhase<SP> {
+		&mut self.phase
+	}
+
+	pub fn phase_take(self) -> ChannelPhase<SP> {
+		self.phase
+	}
```

Comment:
Seems like we could move `context` to `ChannelWrapper` to avoid needing to deconstruct it. Otherwise, we still need to re-insert into the map. So, IIUC, this change doesn't really help us any.

Reply:
Maybe we should discuss this offline, I don't fully get it.
When a pending channel becomes funded, we need to remove&insert because the channel ID changed.
But when only the phase changes, there is no longer need for removal, see `move_v2_to_funded`.
The method to change the phase internally is a bit tricky, as the context owned by the old phase needs to be moved to the new phase.
Moving the context to ChannelWrapper would not solve the cases when there are (will be) more than one channel contexts.

Reply:
> Maybe we should discuss this offline, I don't fully get it. When a pending channel becomes funded, we need to remove&insert because the channel ID changed. But when only the phase changes, there is no longer need for removal, see `move_v2_to_funded`.

Right, for outbound channels the id will change. I had splicing on my mind where we shouldn't change once already funded, IIRC.

> The method to change the phase internally is a bit tricky, as the context owned by the old phase needs to be moved to the new phase. Moving the context to ChannelWrapper would not solve the cases when there are (will be) more than one channel contexts.

That's why I was saying to move `context` into `ChannelWrapper`. Then you can simply override `phase` with a new variant. Could you remind me when we'll have more than one channel context and how they'll be stored? Is that for splicing? Will that involve a phase transition? They could possibly be `swap`ed if contained within a `Vec`, for instance, or `take`n from an internal `Option` if not required.

Reply:
`Funded` -> `RefundingV2` -> `Funded`
In `RefundingV2` there is the original (pre-splice) context, and a pending one (inbound or outbound, V2 negotiating). When that's negotiating, it becomes (Funded)Channel, but still within `RefundingV2` phase. When it is confirmed, the splice is completed, and becomes `Funded`.
With splice RBF variants there can be another pending, and several funding variants.
So in `RefundingV2` there is at least two contexts: pre-splice and post-splice, or even more with RBF.
In places where the `Funded` channel is taken, for `RefundingV2` we sometimes have to take the pre-splice channel (e.g. when listing funded channels, or forwarding a payment), sometimes the post-splice (e.g. when tx negotiation msgs are handled), and sometimes _both_ -- e.g. in commitment update.

Reply:
Moving the context to `ChannelWrapper`: I don't see how this could work. What would be left in the ChannelPhases then?

Reply:
> Moving the context to `ChannelWrapper`: I don't see how this could work. What would be left in the ChannelPhases then?

Not much else for `FundedChannel`, which would mean `RefundingV2` would only need to hold `ChannelVariants` since the `ChannelWrapper` would contain the "current" `ChannelContext`, leaving any other "pending" `ChannelContext`s within the corresponding phase. Then, for splicing, the pending `ChannelContext` can be moved / swapped to `ChannelWrapper` when transitioning to `Funded`. At least that's what I was thinking at a high-level to avoid the `Option`.